sqlglot/optimizer/rules/qualify_columns.py

```diff
@@ -0,0 +1,391 @@
+import dataclasses
+import itertools
+import typing
+
+import sqlglot.expressions as exp
+from sqlglot import parse_one
+from sqlglot.errors import OptimizeError
+
+
+def qualify_columns(expression, schema):
+    """
+    Rewrite sqlglot AST to have fully qualified columns.
+
+    Example:
+        >>> import sqlglot
+        >>> schema = {"tbl": {"col": "INT"}}
+        >>> expression = sqlglot.parse_one("SELECT col FROM tbl")
+        >>> qualify_columns(expression, schema).sql()
+        'SELECT "tbl"."col" AS "col" FROM tbl'
+
+    Args:
+        expression (sqlglot.Expression): expression to qualify
+        schema (dict): Mapping of table names to all available columns
+    Returns:
+        sqlglot.Expression: qualified expression
+    """
+    expression = expression.copy()
+    _qualify_statement(expression, Context(schema=schema))
+    return expression
+
+
+@dataclasses.dataclass
+class Context:
+    """
+    Scoped context for a SELECT statement.
+
+    This is a convenience for passing information between CTEs, subqueries, etc.
+
+    Attributes:
+        schema: Mapping of table names to all available columns
+        sequence: Auto-incrementing sequence
+        selectables: Mapping of name to selectable columns
+        aliased_columns: List of column names.
+            This should be set for derived tables or CTEs where the parent context
+            defines column names for an alias, e.g. "SELECT * FROM (SELECT * FROM x) AS y(a, b)"
+    """
+
+    schema: dict
+    sequence: typing.Iterable = dataclasses.field(default_factory=itertools.count)
+    selectables: dict = dataclasses.field(default_factory=dict)
+    aliased_columns: list = dataclasses.field(default_factory=list)
+
+    def add_selectables(self, *selectables, check_unique=True):
+        return Context(
+            schema=self.schema,
+            sequence=self.sequence,
+            selectables=_merge(
+                self.selectables, *selectables, check_unique=check_unique
+            ),
+            aliased_columns=self.aliased_columns,
+        )
+
+    def branch(self, aliased_columns=None):
+        return Context(
+            schema=self.schema,
+            sequence=self.sequence,
+            selectables=self.selectables,
+            aliased_columns=aliased_columns or [],
+        )
+
+
+def _qualify_statement(expression, context):
+    """
+    Search SELECT or UNION for columns to qualify.
+
+    Args:
+        expression (exp.Select or exp.Union): expression to search
+        context (Context): current context
+    Returns:
+        list: output column names
+    """
+    if isinstance(expression, exp.Select):
+        return _qualify_select(expression, context)
+    if isinstance(expression, exp.Union):
+        left = _qualify_select(expression.this, context)
+        right = _qualify_statement(expression.args.get("expression"), context)
+        if set(left) != set(right):
+            raise OptimizeError("UNION columns not equal")
```

Comment:
in sqlglot, we have different error levels so that users can ignore errors, or just warn, should we do that as well here?

Reply:
Ah, interesting.

That might be tricky to implement here. We'd probably want to move these rules into classes so it's easier to store global context (i.e. a list of error messages). We also might have to choose _how_ to ignore errors on a case by case basis... since doing _nothing_ might lead to other errors later.

I get the value of consistent interfaces though... 
And it might make sense to have better error reporting here. An exception from some nested expression might be tricky to debug.

Reply:
true, let's keep it then. the issue being if a column is ambiguous, we won't be able to do some of the later optimizations. we can keep it